src/data/setoid.lean

```diff
@@ -0,0 +1,488 @@
+  (H : ∀ a, ∃ b ∈ c, a ∈ b ∧ ∀ b' ∈ c, a ∈ b' → b = b')
+  {x b b'} (hc : b ∈ c) (hb : x ∈ b) (hc' : b' ∈ c) (hb' : x ∈ b') :
+  b = b' :=
+let ⟨_, _, _, h⟩ := H x in (h b hc hb).symm.trans $ h b' hc' hb'
+
+/-- Makes an equivalence relation from a set of sets partitioning α. -/
+def mk_classes (c : set (set α))
+  (H : ∀ a, ∃ b ∈ c, a ∈ b ∧ ∀ b' ∈ c, a ∈ b' → b = b') :
+  setoid α :=
+⟨λ x y, ∀ b ∈ c, x ∈ b → y ∈ b, ⟨λ _ _ _ hx, hx,
+ λ x _ h _ hb hy, let ⟨z, hc, hx, hz⟩ := H x in
+    eq_of_mem_eqv_class H hc (h z hc hx) hb hy ▸ hx,
+ λ x y z h1 h2 b hc hb, let ⟨v, hvc, hy, hv⟩ := H y in let ⟨w, hwc, hz, hw⟩ := H z in
+    (eq_of_mem_eqv_class H hwc hz hvc $ h2 v hvc hy).trans
+      (eq_of_mem_eqv_class H hvc hy hc $ h1 b hc hb) ▸ hz⟩⟩
+
+/-- Makes the equivalence classes of an equivalence relation. -/
+def classes (r : setoid α) : set (set α) :=
+{s | ∃ y, s = {x | r.rel x y}}
+
+lemma mem_classes (r : setoid α) (y) : {x | r.rel x y} ∈ r.classes := ⟨y, rfl⟩
+
+/-- Two equivalence relations are equal iff all their equivalence classes are equal. -/
+lemma eq_iff_classes_eq' {r₁ r₂ : setoid α} :
+  r₁ = r₂ ↔ ∀ x, {y | r₁.rel x y} = {y | r₂.rel x y} :=
+⟨λ h x, h ▸ rfl, λ h, ext' $ λ x, (set.ext_iff _ _).1 $ h x⟩
+
+/-- Two equivalence relations are equal iff their equivalence classes are equal. -/
+lemma classes_inj {r₁ r₂ : setoid α} :
+  r₁ = r₂ ↔ r₁.classes = r₂.classes :=
+⟨λ h, h ▸ rfl, λ h, ext' $ λ a b,
+  ⟨λ h1, let ⟨w, hw⟩ := show _ ∈ r₂.classes, by rw ←h; exact r₁.mem_classes a in
+      r₂.trans' (show a ∈ {x | r₂.rel x w}, from hw ▸ r₁.refl' a) $
+        r₂.symm' (show b ∈ {x | r₂.rel x w}, by rw ←hw; exact r₁.symm' h1),
+   λ h1, let ⟨w, hw⟩ := show _ ∈ r₁.classes, by rw h; exact r₂.mem_classes a in
+      r₁.trans' (show a ∈ {x | r₁.rel x w}, from hw ▸ r₂.refl' a) $
+        r₁.symm' (show b ∈ {x | r₁.rel x w}, by rw ←hw; exact r₂.symm' h1)⟩⟩
+
+/-- The empty set is not an equivalence class. -/
+lemma empty_not_mem_classes {r : setoid α} : ∅ ∉ r.classes :=
+λ ⟨y, hy⟩, set.not_mem_empty y $ hy.symm ▸ r.refl' y
+
+/-- Equivalence classes partition the type. -/
+lemma classes_eqv_classes {r : setoid α} :
+  ∀ a, ∃ b ∈ r.classes, a ∈ b ∧ ∀ b' ∈ r.classes, a ∈ b' → b = b' :=
+λ a, ⟨{x | r.rel x a}, r.mem_classes a,
+  ⟨r.refl' a, λ s ⟨y, h⟩ ha, by rw h at *; ext;
+    exact ⟨λ hx, r.trans' hx ha, λ hx, r.trans' hx $ r.symm' ha⟩⟩⟩
+
+/-- If x ∈ α is in 2 equivalence classes, the equivalence classes are equal. -/
+lemma eq_of_mem_classes {r : setoid α} {x b} (hc : b ∈ r.classes)
+  (hb : x ∈ b) {b'} (hc' : b' ∈ r.classes) (hb' : x ∈ b') : b = b' :=
+eq_of_mem_eqv_class classes_eqv_classes hc hb hc' hb'
+
+/-- The elements of a set of sets partitioning α are the equivalence classes of the
+    equivalence relation defined by the set of sets. -/
+lemma eq_eqv_class_of_mem {c : set (set α)}
+  (H : ∀ a, ∃ b ∈ c, a ∈ b ∧ ∀ b' ∈ c, a ∈ b' → b = b')
+  {s y} (hs : s ∈ c) (hy : y ∈ s) : s = {x | rel (mk_classes c H) x y} :=
```

Comment:
Maybe `(mk_classes c H).rel x y` reads more naturally, here and in a few places later.